### ISSUE TITLE
On windows, calling disk.DiskIOCounters() causes WMI service to disrupt write activity in elasticsearch

### ISSUE THREAD

**Issue by oldenbur** [shirou/gopsutil#124]
Our application wraps an elasticsearch instance which is constantly indexing (writing) while periodically collecting stats using the following gopsutil calls:

``` go
cpu.CPUPercent(0, true)
disk.DiskIOCounters()
mem.VirtualMemory()
disk.DiskPartitions(true)
disk.DiskUsage(...)
```

We noticed using jstack that every 15 minutes, elasticsearch writes were blocking for a short window, but long enough to fall behind and be disruptive.  We noticed that stopping and disabling the WMI service made the problem disappear.  Ultimately, we found that not calling the `disk.DiskIOCounters()` method fixed the problem while the WMI process was running, suggesting that method was doing something to cause the disruptive WMI behavior.

**Comment by shirou** [comment]
`disk.DiskIOCounters()` uses 
- https://github.com/StackExchange/wmi

and wmi uses

- https://github.com/go-ole/go-ole

From gopsutil, just invoke `wmi.Query`. I updated latest go-ole in my environment, DiskIOCounter WMI Query returns error. I will dig it deeper, but please wait until this weekend.

**Comment by shirou** [comment]
When I test now, I can not reproduce that error, with same code, same PC. Perhaps, at the last time the WMI service was something bad, I got the error. But something fixed.

`disk.DiskIOCounters()` always do something bad with latest go-ole and WMI? If so, could you send me some more detail? 

Thank you.

**Comment by oldenbur** [comment]
I updated the following to the latest versions:
 * github.com/StackExchange/wmi
 * github.com/go-ole/go-ole
 * github.com/shirou/gopsutil

I am still observing the problem on multiple systems.  The operating system is 64bit Windows Server 2012 R2 Standard.  The machine is a Dell Server PER730XD with 2 wGHz Intel Xeon CPUs and 128GB RAM.  I included a grafana screenshot showing the outages in CPU and Memory.  These outages go away if I don't make any calls to `disk.DiskIOCounters()` as I mentioned.  We have other ways to get this information, so this is not a serious problem for us, just letting you know.  Thanks for your attention and please let me know if I can test anything for you.

![screen shot 2015-12-15 at 8 27 09 am](https://cloud.githubusercontent.com/assets/4369022/11814734/b19f6956-a305-11e5-9904-8741572bb61d.png)

**Comment by shirou** [comment]
Thanks for sharing information. I will try on Windows Server 2012 R2. And, if there is no problem, could you tell me how to get disk IO Counter with another way? I am not cleave to use WMI. If there are better way, I want to change to.

**Comment by ihocho** [comment]
I am using gopsutil to get system information. I can get most information except process's memory usage on windows.
-----------------------------------------------------------
proc, err := process.NewProcess(value)
mem, err := proc.MemoryInfo()
mem.RSS <-- "Error : not implemented yet"
-----------------------------------------------------------
If i am not able to get that now, when or how to get it?

**Comment by sparrc** [comment]
@shirou, if it helps, in Telegraf we use win perf mon queries to get most system stats, you can see examples and the code here: https://github.com/influxdata/telegraf/tree/master/plugins/inputs/win_perf_counters

**Comment by shirou** [comment]
@sparrc, thank you for the information. But, it seems using `github.com/lxn/win` and that use cgo, right?

I know the `PdhOpenQuery` approach like [this](https://github.com/mackerelio/mackerel-agent/tree/master/metrics/windows). However, I am not familiar with Windows, I could not understand, so I change to use WMI.

But as you suggested, we have to change to use PDH, though.

**Comment by Lomanic** [comment]
Small update: telegraf doesn't use github.com/lxn/win anymore and this part of the code doesn't rely on cgo.

**Comment by Lomanic** [comment]
Closed by #850
